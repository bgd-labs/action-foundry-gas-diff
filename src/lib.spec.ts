import { expect, it, describe } from "vitest";
import { formatDiffMd, snapshotDiff } from "./lib";

const before = {
  "test_removed": "123",
  "test_unchanged": "456",
  "test_bigger": "789",
  "test_smaller": "789",
}

const after = {
  "test_unchanged": "456",
  "test_bigger": "1111",
  "test_smaller": "654",
  "test_added": "789",
}

describe("lib", () => {
  it("should throw when not passed numbers in before", () => {
    expect(() => snapshotDiff({
      before: {
        "test": "abc",
      },
      after: {
        "test": "123",
      },
    })).toThrowErrorMatchingInlineSnapshot(`[Error: The following keys in before are not numbers: test]`);
  })

  it("should throw when not passed numbers in after", () => {
    expect(() => snapshotDiff({
      before: {
        "test": "123",
      },
      after: {
        "test": "abc",
      },
    })).toThrowErrorMatchingInlineSnapshot(`[Error: The following keys in after are not numbers: test]`);
  })

  it("should detect differences between two different snapshots", () => {


    expect(
      snapshotDiff({
        before,
        after
      }),
    ).toMatchInlineSnapshot(`
      {
        "added": {
          "test_added": "789",
        },
        "changed": {
          "test_bigger": "<sup>‚Üë41% (+322)</sup> 1,111",
          "test_smaller": "<sup>‚Üì17% (-135)</sup> 654",
        },
        "removed": {
          "test_removed": "123",
        },
        "unchanged": {
          "test_unchanged": "456",
        },
      }
    `);
  });

  it("nicely formats the diff as markdown", () => {
    const diffA = snapshotDiff({
      before,
      after
    })

    const diffB = snapshotDiff({
      before,
      after
    })

    const result = formatDiffMd("Abcdef", [{ path: "path_a", diff: diffA }, { path: "path_b", diff: diffB }])
    console.log(result)
    expect(result).toMatchInlineSnapshot(`
      "### ‚ôªÔ∏è Abcdef
      | Path | Value |
      | --- | ---: |
      | **path_a** |  |
      | test_bigger | <sup>‚Üë41% (+322)</sup> 1,111 |
      | test_smaller | <sup>‚Üì17% (-135)</sup> 654 |
      | test_removed | 123 |
      | test_added | 789 |
      | **path_b** |  |
      | test_bigger | <sup>‚Üë41% (+322)</sup> 1,111 |
      | test_smaller | <sup>‚Üì17% (-135)</sup> 654 |
      | test_removed | 123 |
      | test_added | 789 |

      <details><summary>üîï Unchanged</summary>

      | Path | Value |
      | --- | ---: |
      | **path_a** |  |
      | test_unchanged | 456 |
      | **path_b** |  |
      | test_unchanged | 456 |
      </details>"
    `)
  })
});
