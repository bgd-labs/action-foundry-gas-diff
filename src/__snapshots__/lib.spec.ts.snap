// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lib > should generate a well formatted report with empty root 1`] = `
"### [Create3Factory](https://github.com/src/contracts/create3/Create3Factory.sol)

- gas: 224263
| Method | min | mean | median | max | calls |
| --- | ---: | ---: | ---: | ---: | ---: |
create(bytes32,bytes) | 286388 | 296666 | 287062 | 397068 | 6 |
predictAddress(address,bytes32) | 1024 | 1024 | 1024 | 1024 | 6 |


### [ProxyAdmin](https://github.com/src/contracts/transparent-proxy/ProxyAdmin.sol)

- gas: 0
| Method | min | mean | median | max | calls |
| --- | ---: | ---: | ---: | ---: | ---: |
owner() | 386 | 386 | 386 | 386 | 6 |


### [TransparentProxyFactory](https://github.com/src/contracts/transparent-proxy/TransparentProxyFactory.sol)

- gas: 1609662
| Method | min | mean | median | max | calls |
| --- | ---: | ---: | ---: | ---: | ---: |
createDeterministic(address,address,bytes,bytes32) | 574115 | 534507 | 574589 | 574703 | 6 |
createDeterministicProxyAdmin(address,bytes32) | 421268 | 421658 | 421850 | 421856 | 6 |
predictCreateDeterministic(address,address,bytes,bytes32) | 12057 | 12057 | 12057 | 12057 | 6 |
predictCreateDeterministicProxyAdmin(bytes32) | 6236 | 6236 | 6236 | 6236 | 6 |


### [TransparentUpgradeableProxy](https://github.com/src/contracts/transparent-proxy/TransparentUpgradeableProxy.sol)

- gas: 0
| Method | min | mean | median | max | calls |
| --- | ---: | ---: | ---: | ---: | ---: |
getFoo() | 1042 | 1042 | 1042 | 1042 | 6 |


### [ImplOwnableWithGuardian](https://github.com/test/OwnableWithGuardian.t.sol)

- gas: 285211
| Method | min | mean | median | max | calls |
| --- | ---: | ---: | ---: | ---: | ---: |
guardian() | 2286 | 2286 | 2286 | 2286 | 1 |
owner() | 2341 | 2341 | 2341 | 2341 | 1 |
updateGuardian(address) | 25944 | 28245 | 28283 | 30394 | 6 |


### [TestPermissionlessRescuable](https://github.com/test/PermissionlessRescuable.t.sol)

- gas: 879072
| Method | min | mean | median | max | calls |
| --- | ---: | ---: | ---: | ---: | ---: |
emergencyEtherTransfer(uint256) | 59057 | 59900 | 59900 | 60743 | 2 |
emergencyTokenTransfer(address,uint256) | 34213 | 50627 | 51185 | 65925 | 4 |
receive() | 0 | 0 | 0 | 0 | 1 |
whoShouldReceiveFunds() | 2566 | 2566 | 2566 | 2566 | 1 |


"
`;

exports[`lib > should generate a well formatted report with existing root & skip unchanged 1`] = `
"### [Create3Factory](https://github.com/src/contracts/create3/Create3Factory.sol)

- gas: 224263
| Method | min | mean | median | max | calls |
| --- | ---: | ---: | ---: | ---: | ---: |
create(bytes32,bytes) | <sup title="296388">↓-3.4%</sup>286388 | 296666 | 287062 | <sup title="297068">↑34%</sup>397068 | 6 |


### [TransparentProxyFactory](https://github.com/src/contracts/transparent-proxy/TransparentProxyFactory.sol)

- gas: 1609662
"
`;

exports[`lib > should generate a well formatted report with existing root 1`] = `
"### [Create3Factory](https://github.com/src/contracts/create3/Create3Factory.sol)

- gas: 224263
| Method | min | mean | median | max | calls |
| --- | ---: | ---: | ---: | ---: | ---: |
create(bytes32,bytes) | <sup title="296388">↓-3.4%</sup>286388 | 296666 | 287062 | <sup title="297068">↑34%</sup>397068 | 6 |
predictAddress(address,bytes32) | 1024 | 1024 | 1024 | 1024 | 6 |


### [ProxyAdmin](https://github.com/src/contracts/transparent-proxy/ProxyAdmin.sol)

- gas: 0
| Method | min | mean | median | max | calls |
| --- | ---: | ---: | ---: | ---: | ---: |
owner() | 386 | 386 | 386 | 386 | 6 |


### [TransparentProxyFactory](https://github.com/src/contracts/transparent-proxy/TransparentProxyFactory.sol)

- gas: 1609662
| Method | min | mean | median | max | calls |
| --- | ---: | ---: | ---: | ---: | ---: |
createDeterministic(address,address,bytes,bytes32) | 574115 | 534507 | 574589 | 574703 | 6 |
createDeterministicProxyAdmin(address,bytes32) | 421268 | 421658 | 421850 | 421856 | 6 |
predictCreateDeterministic(address,address,bytes,bytes32) | 12057 | 12057 | 12057 | 12057 | 6 |
predictCreateDeterministicProxyAdmin(bytes32) | 6236 | 6236 | 6236 | 6236 | 6 |


### [TransparentUpgradeableProxy](https://github.com/src/contracts/transparent-proxy/TransparentUpgradeableProxy.sol)

- gas: 0
| Method | min | mean | median | max | calls |
| --- | ---: | ---: | ---: | ---: | ---: |
getFoo() | 1042 | 1042 | 1042 | 1042 | 6 |


### [ImplOwnableWithGuardian](https://github.com/test/OwnableWithGuardian.t.sol)

- gas: 285211
| Method | min | mean | median | max | calls |
| --- | ---: | ---: | ---: | ---: | ---: |
guardian() | 2286 | 2286 | 2286 | 2286 | 1 |
owner() | 2341 | 2341 | 2341 | 2341 | 1 |
updateGuardian(address) | 25944 | 28245 | 28283 | 30394 | 6 |


### [TestPermissionlessRescuable](https://github.com/test/PermissionlessRescuable.t.sol)

- gas: 879072
| Method | min | mean | median | max | calls |
| --- | ---: | ---: | ---: | ---: | ---: |
emergencyEtherTransfer(uint256) | 59057 | 59900 | 59900 | 60743 | 2 |
emergencyTokenTransfer(address,uint256) | 34213 | 50627 | 51185 | 65925 | 4 |
receive() | 0 | 0 | 0 | 0 | 1 |
whoShouldReceiveFunds() | 2566 | 2566 | 2566 | 2566 | 1 |


"
`;
